### PR TITLE
Skyline End to end testing

### PR DESCRIPTION
**Pull request by michaelshin**
Wrote a script to run baseline measurements of certain example models, then ran Skyline to profile the same the measurements and to compare the results.

### CODE REVIEW

**Review by anurlybayev — COMMENTED** [review]
Can you add testing instructions to the PR description? Few of the parameters that we can pass along to the script to see the results?

**Comment by michaelshin** [comment]
To run this test, a user must take the following steps

1. Build Skyline in a virtual environment in the parent directory of this current repository.
2. Change directory to skyline/test/e2e_test.
3. Run the script by calling `python e2e_test.py`
5. The results will be found under skyline/test/e2e_test/results with the following format: **ISO date format-(GPU name).json** where the date and GPU name are filled out accordingly.

To modify the list of tests run, you can update config.json which is found under skyline/test/e2e_test. Note that the intention is for these scripts to be run in cloud instances with different GPU types to act as a sanity check for Skyline's profiling.

**Review by anurlybayev — CHANGES_REQUESTED** [review]
So the branch was broken for me due to the issue described in #17. After fixing the issue I ran:
```
$ poetry install
$ poetry run python e2e_test.py
Traceback (most recent call last):
  File "e2e_test.py", line 14, in <module>
    from utils import BackendContext, SkylineSession
  File "/home/anurlybayev/Developer/CentML/skyline/test/e2e_test/./../utils.py", line 8, in <module>
    import pytest
ModuleNotFoundError: No module named 'pytest'

$ poetry add pytest
Using version ^7.2.0 for pytest

Updating dependencies
Resolving dependencies... (1.1s)

Writing lock file

Package operations: 7 installs, 1 update, 0 removals

  • Installing pyparsing (3.0.9)
  • Installing exceptiongroup (1.0.0)
  • Installing iniconfig (1.1.1)
  • Installing packaging (21.3)
  • Installing pluggy (1.0.0)
  • Updating pyrsistent (0.19.1 /home/anurlybayev/.cache/pypoetry/artifacts/fc/23/1c/1530f80f8832b64108c07f7eb2ff390c5c6031e7d6b72af68aebbc95a8/pyrsistent-0.19.1-py3-none-any.whl -> 0.19.1)
  • Installing tomli (2.0.1)
  • Installing pytest (7.2.0)

$ poetry run python e2e_test.py
Traceback (most recent call last):
  File "e2e_test.py", line 14, in <module>
    from utils import BackendContext, SkylineSession
  File "/home/anurlybayev/Developer/CentML/skyline/test/e2e_test/./../utils.py", line 10, in <module>
    from termcolor import colored
ModuleNotFoundError: No module named 'termcolor'

$ poetry add termcolor
Using version ^2.1.0 for termcolor

Updating dependencies
Resolving dependencies... (0.2s)

Writing lock file

Package operations: 1 install, 0 updates, 0 removals

  • Installing termcolor (2.1.0)

$ poetry run python e2e_test.py
/home/anurlybayev/Developer/CentML/skyline/skyline
/home/anurlybayev/Developer/CentML/skyline/skyline/pyproject.toml
/home/anurlybayev/Developer/CentML/skyline/.venv/lib/python3.8/site-packages/torch/cuda/__init__.py:146: UserWarning:
NVIDIA GeForce RTX 3060 Ti with CUDA capability sm_86 is not compatible with the current PyTorch installation.
The current PyTorch install supports CUDA capabilities sm_37 sm_50 sm_60 sm_70.
If you want to use the NVIDIA GeForce RTX 3060 Ti GPU with PyTorch, please check the instructions at https://pytorch.org/get-started/locally/

  warnings.warn(incompatible_device_warn.format(device_name, capability, " ".join(arch_list), device_name))
/home/anurlybayev/Developer/CentML/skyline/skyline
/home/anurlybayev/Developer/CentML/skyline/skyline/pyproject.toml
Traceback (most recent call last):
  File "baseline_test_runner.py", line 8, in <module>
    import skyline.analysis.session as session
  File "/home/anurlybayev/Developer/CentML/skyline/skyline/analysis/session.py", line 19, in <module>
    import habitat
ModuleNotFoundError: No module named 'habitat'
Traceback (most recent call last):
  File "e2e_test.py", line 153, in <module>
    test_model_metrics_json_dump()
  File "e2e_test.py", line 134, in test_model_metrics_json_dump
    model_metrics_runner = ModelMetricsRunner(model_names[i], skyline_bin, entry_points[i])
  File "e2e_test.py", line 53, in __init__
    self.model_metrics = self._get_model_metrics()
  File "e2e_test.py", line 109, in _get_model_metrics
    given_batch_size, baseline_metrics = self._get_baseline_values()
  File "e2e_test.py", line 64, in _get_baseline_values
    with open(file_name, 'rb') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'resnet.pkl'
```

After resolving two missing dependencies issues, see above, I still wasn't able to run the test, due to what seems to be a hard dependency on `habitat`. In short I was not able to follow the testing instructions of this PR.

File: test/e2e_test/baseline_test_runner.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                'iterator': self._iteration_provider(self._model_provider())
+            }
+        )
+        timer.timeit(100)
+        peak_memory_usage = cuda.max_memory_allocated()
+        cuda.empty_cache()
+        max_memory = cuda.mem_get_info()[1]
+        max_batch_size = self._given_batch_size*max_memory/peak_memory_usage
+        # TODO mshin: figure out a way to dynamically modify the divisors of the values
+        return [int(max_batch_size/12), int(max_batch_size/6), int(max_batch_size/3), int(max_batch_size/2)]
```
Comment (ssaini4) [inline test/e2e_test/baseline_test_runner.py:91]: What are the magic numbers for?

File: test/e2e_test/baseline_test_runner.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            cuda.reset_max_memory_allocated()
+
+            timer = benchmark.Timer(
+                stmt='iterator(*inputs)',
+                globals={
+                    'inputs': self._input_provider(batch_size=iter_batchsize),
+                    'iterator': self._iteration_provider(self._model_provider())
+                }
+            )
+            measurement = timer.timeit(100)
```
Comment (ssaini4) [inline test/e2e_test/baseline_test_runner.py:106]: Same as above. Maybe a comment about what `100` is  here

File: test/e2e_test/baseline_test_runner.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        description="Baseline test runner: Gets profiling metrics for a given metric."
+                    "Used to compare Skyline's profiling results",
+    )
+    parser.add_argument(
+        "name",
+        help="Name of the model that we are profiling. Used for naming the export file",
+    )
+    parser.add_argument(
+        "entry_point",
+
```
Comment (ssaini4) [inline test/e2e_test/baseline_test_runner.py:132]: extra line

File: test/e2e_test/e2e_test.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+from jsonschema import validate
+from math import nan
+import numpy as np
+import os
+import pickle
+import torch
+from typing import List, Tuple
+import subprocess
+import sys
+sys.path.append('./..')
```
Comment (ssaini4) [inline test/e2e_test/e2e_test.py:13]: This should be after all the imports, right?

File: test/e2e_test/e2e_test.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+def _update_entry_point_batch_size(file_name, original_batch_size, target_batch_size):
+    if original_batch_size == target_batch_size:
+        pass
+    
+    with open(file_name, 'r') as file :
+        filedata = file.read()
+
+    # ensure that batch_size is a parameter
+    assert(filedata.find("batch_size=") != -1)
```
Comment (ssaini4) [inline test/e2e_test/e2e_test.py:38]: This is going to bubble up to the user as an AssertionError. Should we add a try..except clause and throw something more user friendly?

File: test/e2e_test/e2e_test.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        if type(result) == subprocess.CompletedProcess:
+            file_name = self.model_name + '.pkl'
+            baseline_profiling_metrics_list = []
+            given_batch_size = nan
+            with open(file_name, 'rb') as f:
+                baseline_metrics = pickle.load(f)
+                # some sanity checks
+                assert(type(baseline_metrics) == e2e_utils.BaselineMetrics)
+                assert(baseline_metrics.model_name == self.model_name)
+                assert(baseline_metrics.entry_point == self._entry_point)
```
Comment (ssaini4) [inline test/e2e_test/e2e_test.py:69]: I would recommend refactoring this into a separate method

File: test/e2e_test/baseline_test_runner.py (hunk 1 above)
Comment (michaelshin) [reply test/e2e_test/baseline_test_runner.py:91]: They're a hacky way to get a subset of usable batch sizes for a given model. I don't think this makes sense to use anymore and it's better to specify a list of batch sizes for each model

File: test/e2e_test/baseline_test_runner.py (hunk 2 above)
Comment (michaelshin) [reply test/e2e_test/baseline_test_runner.py:106]: Good point. The 100 here signifies the number of times to run this measurement so we can average it out

File: test/e2e_test/e2e_test.py (hunk 4 above)
Comment (michaelshin) [reply test/e2e_test/e2e_test.py:13]: This was to ensure the utils module can be imported. I am planning to refactor this test completely so I will get rid of that line

File: test/e2e_test/e2e_test.py (hunk 5 above)
Comment (michaelshin) [reply test/e2e_test/e2e_test.py:38]: Good point I will update this

**Comment by michaelshin** [comment]
@anurlybayev I rebuilt Skyline in a clean new virtual environment and it still has a hard dependency on `habitat`. @jimgao1 We should update Skyline so it doesn't have this dependency. 

Regardless I am planning to move this test to a new repo solely for end to end testing instead. I think this approach makes more sense as the fully fledged version will test both habitat and skyline

**Pull request closed without merging**